### ISSUE TITLE
Events do not show up on Windows 10 Mobile - bug in WP8.1 reappears - OC8.1 & higher

### ISSUE THREAD

**Issue by elvenbone** [owncloudarchive/calendar#966]
This is exactly the same issue as described and fixed for Windows Phone version 8.1 in:
https://github.com/owncloud/calendar/issues/853
https://github.com/fruux/sabre-dav/commit/9bd444392d57e55348b8edb8d1c8a0d8f89bb47d
Difference here is the operating system - the WP8 bug reappears on Windows 10 Mobile.

**--> I already found a solution, but am not sure how to proceed. Please scroll down below the standard bug description for details.**

### Steps to reproduce
1. Add CalDAV account to Windows 10 Mobile accounts by using the iCloud account and changing the CalDAV address (This is a well-known workaround already known in WP8 and WP8.1)
2. Save the account and wait a bit until you assume synchronisation is finished.
3. Open any calendar app on your phone.

### Expected behaviour
- The owncloud calendars should appear in the calendar app
- Their events should appear in the calendar app

### Actual behaviour
- The owncloud calendars appear in the calendar app
- No errors arise, calendars sync just fine
- BUT: No events show up, the calendars (seemingly) remain empty!

### Server configuration
Owncloud 8.1.x and owncloud 8.2,
relevant to all server platforms and their configuration.

### Client configuration
Smartphone with Windows 10 Mobile, bug any calendar app on it.
Note: W10M phones will be in stores starting next month, but users can download it on their current phones much like MS did with the Win10 desktop version "insider" preview.

The rest of the issue template should be irrelevant here, necessary details are described in the following solution/workaround.

### SOLUTION / WORKAROUND

Windows 10 Mobile sends a **different user agent** (`MSFT-WIN-4` -- don't ask me what that 4 means) than Windows Phone 8 (`MSFT-WP`), according to my nginx server's access.log.

I suggest we enhance the workaround implemented in sabre-dav 2.1.6 (see below and 2nd link at the top):
```php
if ($node instanceof ICalendarObjectContainer && $depth === 0) {

            if(strpos($this->server->httpRequest->getHeader('User-Agent'), 'MSFT-WP/') === 0) {
                // Windows phone incorrectly supplied depth as 0, when it actually
                // should have set depth to 1. We're implementing a workaround here
                // to deal with this.
                $depth = 1;
            } else {
                throw new BadRequest('A calendar-query REPORT on a calendar with a Depth: 0 is undefined. Set depth to 1.');
            }

        }
```
...by adding a clause such as this:
```php
if ($node instanceof ICalendarObjectContainer && $depth === 0) {

            if(strpos($this->server->httpRequest->getHeader('User-Agent'), 'MSFT-WP/') === 0) {
                // Windows phone incorrectly supplied depth as 0, when it actually
                // should have set depth to 1. We're implementing a workaround here
                // to deal with this.
                $depth = 1;
            } elseif (strpos($this->server->httpRequest->getHeader('User-Agent'), 'MSFT-WIN-4/') === 0) {
                // The same bug exists in Windows 10 Mobile (the WP8.1 successor), but W10M has a different user-agent.
                $depth = 1;
            } else {
                throw new BadRequest('A calendar-query REPORT on a calendar with a Depth: 0 is undefined. Set depth to 1.');
            }

        }
```

### Remarks:
- This workaround works immediately after editing the file when syncing the phone calendar again.
- I do not know anything about PHP programming, please feel free to implement it differently if there is a cleaner or more performant way to do so!

As you can see, we may need to fix this upstream in sabre-dav again, thanks to Microsoft not implementing standards properly...

I am using github for the first time now and have never used such a system, thus I do not know how to proceed from here, where to create a pull-request or commit (here or directly on sabre-dav or both?), how a fixed sabre-dav comes into owncloud etc...
I can later try to post it on sabre-dav, but if you have routine in this, I'll be glad if you do so :)

Hope this "bug" report contains everything necessary, long logs or config.php should not be needed here due to the known nature of the problem (see first link on top).

best regards,
elvenbone

**Comment by RealRancor** [comment]
There is a similar report to upstream available here: https://github.com/fruux/sabre-dav/issues/727

I think its the best to report this also there as this is AFAIK something which should be fixed upstream (in SabreDAV) an not within oC itself. Or am i wrong?

**Comment by RealRancor** [comment]
@elvenbone I think you could close here as it got fixed with https://github.com/fruux/sabre-dav/commit/1feeb1bb87c5176ab1c9886c0761c7ac2ef151c4. Furthermore https://github.com/owncloud/calendar/issues/954 is still open which is tracking the oC part.

**Comment by slevit** [comment]
Same behavior occurred after Windows 10 *DESKTOP* fall update to build 10586. Solved with elvenbone's solution substituting MSFT-WIN-4/ with MSFT-WIN-3/.

**Comment by RealRancor** [comment]
@slevit As you can see in the posted links above this is already fixed upstream. I think this should be closed and a new issue in core should be created to request an update to the SabreDAV version shipping this fix.

**Comment by slevit** [comment]
@RealRancor I'm not very well versed in the protocol here, so please bear with me. I just reviewed the upstream fix and I see what you're saying. However, it looks to me like it covers all flavors of Win10 and a new issue created in core shouldn't be necessary. It's just a question of whether or not it's going to show up in the next interim release of ownCloud. Until that happens, the patch I made to my own installation temporarily restores the functionality I lost after the 10586 update. For what it's worth, there was no problem before the 10586 update was applied.

**Comment by RealRancor** [comment]
@slevit This fix won't be shipped that fast in oC. Upgrades of SabreDAV doesn't happen that often so you probably need to patch that after every oC update until SabreDAV 3.1 (which includes this fix according to the changelog) will be included sometime.

**Comment by xChr1s** [comment]
Interestingly, with the patch everything works fine except one thing which doesn't produce any error:
If I add events which are 'far' in the future, in my example august, the events appear shortly and then disappear in the calendar app on mobile and win 10. However, events are listed in the owncloud calendar but do not show up in the apps. There is also nothing in the error logs. Can anyone reproduce this?
OC: 8.2.2
Greetings,
Chris

**Comment by may-bee** [comment]
@xChr1s: I have several users reporting the same issue with their win8 phones. I had thought about any reason for it, but could not come to a sufficient conclusion. 
I will try the provided patch and give it a try soon.
Further investigation needed here.

**Comment by elvenbone** [comment]
@xChr1s, @may-bee: I am convinced this is an issue that neither owncloud nor sabredav are able to handle. I remember reading it affects all calendars on WP8/8.1 and W10M, i.e. Google, Apple, Outlook, etc. as calendar servers as well.
The default calendar app just does not sync more than a few months to keep syncing timespans low (as the dumbest imaginable user won't understand why he should have to wait a few seconds when opening the calendar app).
It seems like a typical case of "this is a feature, not a bug": apparently Microsoft could not be bothered to implement a better background synchronization agent on their mobile system...

You may be able to workaround this issue by using other calendar apps like "TrueCalendar 8" (which also works on W10M I believe), where you can configure how many months in the future and from the past should be loaded from the calendar source/server.

---

On topic: I saw fruux released sabre 3.1 recently, which should include the fix. Hoping owncloud will adopt it soon (unless of course that's too risky compared to staying on 3.0 a little longer) :)

**Comment by may-bee** [comment]
relates to owncloud/core#22724.

**Comment by RealRancor** [comment]
@may-bee Yes, and fixed with the upcoming 9.0.2 which includes an updated SabreDAV fixing this issue: https://github.com/owncloud/core/pull/23839
